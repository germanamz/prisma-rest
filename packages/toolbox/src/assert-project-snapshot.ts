import { Project } from 'ts-morph';

export const assertProjectSnapshot = async (project: Project) => {
  project.getSourceFiles().forEach((sourceFile) => {
    expect(sourceFile.getText()).toMatchSnapshot();
  });
};
