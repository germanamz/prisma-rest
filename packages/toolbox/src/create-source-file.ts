import { Project } from 'ts-morph';
import { normalizeFilename } from './normalize-filename';

export type CreateSourceFileOptions = {
  project: Project;
  name: string;
  dir: string;
  ext?: string;
};

export const createSourceFile = (options: CreateSourceFileOptions) => {
  const {
    dir,
    project,
    name,
    ext = '.ts',
  } = options;

  return project.createSourceFile(
    `${dir}/${normalizeFilename(name)}${ext}`,
    undefined,
    { overwrite: true },
  );
};
