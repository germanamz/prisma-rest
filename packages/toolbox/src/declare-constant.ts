import { SourceFile, VariableDeclarationKind, WriterFunction } from 'ts-morph';
import { Registry } from './registry';

export type DeclareConstantOptions = {
  sourceFile: SourceFile;
  name: string;
  registry: Registry;
  isExported?: boolean;
  initializer: string | WriterFunction;
};

export const declareConstant = ({
  name, sourceFile, initializer, isExported, registry,
}: DeclareConstantOptions) => {
  const statement = sourceFile.addVariableStatement({
    declarationKind: VariableDeclarationKind.Const,
    isExported,
    declarations: [
      {
        name,
        initializer,
      },
    ],
  });

  registry.set(name, sourceFile);

  return statement;
};
