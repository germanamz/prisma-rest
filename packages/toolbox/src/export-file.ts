import { SourceFile } from 'ts-morph';
import path from 'node:path';

export type ExportFileOptions = {
  from: SourceFile;
  target: SourceFile;
};

export const exportFile = ({ from, target }: ExportFileOptions) => {
  const isIndex = target.getBaseName().includes('index');
  let importPath = path.relative(path.dirname(from.getFilePath()), target.getFilePath());

  if (!importPath.startsWith('../')) {
    importPath = `./${importPath}`;
  }

  if (isIndex) {
    importPath = path.dirname(importPath);
  }

  from.addExportDeclaration({
    moduleSpecifier: importPath.replace(RegExp(`${path.extname(target.getFilePath())}$`), ''),
  });
};
