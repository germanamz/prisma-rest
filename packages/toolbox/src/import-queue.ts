import { SourceFile } from 'ts-morph';
import path from 'path';
import { Registry } from './registry';

/**
 * A map of source files to a set of import identifiers.
 */
export type ImportQueue = Map<SourceFile, Set<string>>;

export const addToImportQueue = (q: ImportQueue, sourceFile: SourceFile, identifiers: string[]) => {
  const identifierSet = q.get(sourceFile) || new Set<string>();

  identifiers.forEach((identifier) => identifierSet.add(identifier));

  q.set(sourceFile, identifierSet);
};

export const executeImportQueue = (q: ImportQueue, registry: Registry) => {
  for (const [sourceFile, imports] of q.entries()) {
    for (const item of imports) {
      // TODO: Group imports
      // Import path is './' when directories are the same
      const moduleSpecifier = sourceFile.getDirectoryPath() === registry.get(item)!.getDirectoryPath()
        ? `./${registry.get(item)!.getBaseNameWithoutExtension()}`
        : path.relative(sourceFile.getDirectoryPath(), registry.get(item)!.getDirectoryPath());

      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: [item],
      });
    }
  }
};
