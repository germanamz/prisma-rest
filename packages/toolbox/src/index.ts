export * from './registry';
export * from './import-queue';
export * from './generate-namespace';
