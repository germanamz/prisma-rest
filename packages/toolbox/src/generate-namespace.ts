import { Project, SourceFile } from 'ts-morph';
import { ImportQueue } from './import-queue';
import { Registry } from './registry';


export type GenerateNamespaceHandlerOptions<T> = {
  item: T,
  dir: string;
  project: Project;
  importQueue: ImportQueue;
  registry: Registry;
};

export type GenerateNamespaceOptions<T extends readonly any[]> = {
  dir: string;
  project: Project;
  registry: Registry;
  importQueue: ImportQueue;
  items?: T;
  handler?: (options: GenerateNamespaceHandlerOptions<T[number]>) => SourceFile;
  generator?: (options: GenerateNamespaceHandlerOptions<null>) => SourceFile[];
};

export const generateNamespace = <T extends readonly any[]>({
  dir,
  project,
  items,
  handler,
  registry,
  importQueue,
  generator,
}: GenerateNamespaceOptions<T>) => {
  const file = project.createSourceFile(`${dir}/index.ts`, undefined, { overwrite: true });

  if (generator) {
    const sourceFiles = generator({
      item: null,
      dir,
      project,
      importQueue,
      registry,
    });

    sourceFiles?.forEach((sourceFile) => {
      file.addExportDeclaration({
        moduleSpecifier: `./${sourceFile.getBaseNameWithoutExtension()}`,
      });
    });
  }

  if (handler) {
    items?.forEach((item) => {
      const sourceFile = handler({ item, dir, project, importQueue, registry });

      file.addExportDeclaration({
        moduleSpecifier: `./${sourceFile.getBaseNameWithoutExtension()}`,
      });
    });
  }

  return file;
};
