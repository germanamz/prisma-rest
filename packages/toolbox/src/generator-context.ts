import { Project } from 'ts-morph';
import { DMMF } from '@prisma/generator-helper';
import { MarshalDocument } from '@germanamz/prisma-rest-marshal';
import { ImportQueue } from './import-queue';
import { Registry } from './registry';

// TODO: Remove deprecated dmmf
export type GeneratorContext = {
  dir: string;
  project: Project;
  // @deprecated use marshalDocument instead, present for backwards compatibility
  dmmf: DMMF.Document;
  importQueue: ImportQueue;
  registry: Registry;
  marshalDocument: MarshalDocument;
};

export type MakeGeneratorContextOptions = Partial<Omit<GeneratorContext, 'dmmf'>> & Pick<GeneratorContext, 'dmmf'> & {
  tsConfigFilePath?: string;
};

export const makeGeneratorContext = (options: MakeGeneratorContextOptions): GeneratorContext => ({
  dir: options.dir ?? process.cwd(),
  project: options.project ?? new Project({
    ...(options.tsConfigFilePath ? { tsConfigFilePath: options.tsConfigFilePath } : {}),
    skipAddingFilesFromTsConfig: true,
  }),
  importQueue: options.importQueue ?? new Map() as ImportQueue,
  registry: options.registry ?? new Map() as Registry,
  marshalDocument: options.marshalDocument ?? new MarshalDocument(options.dmmf),
  dmmf: options.dmmf,
});
