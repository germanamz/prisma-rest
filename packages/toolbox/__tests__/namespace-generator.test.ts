import { StatementStructures, StructureKind } from 'ts-morph';
import { assertProjectSnapshot, getMockDir, makeMockProject } from 'test-lib';
import { createSourceFile, namespaceGenerator } from '../src';

describe('namespaceGenerator', () => {
  it('should import the generated files correctly', () => {
    const project = makeMockProject();
    const indexFile = namespaceGenerator({
      project,
      dir: getMockDir(),
      generator: () => [
        createSourceFile({
          project,
          dir: getMockDir(),
          name: 'test',
          content: 'export const test = 1;',
        }),
        undefined,
      ],
    })!;
    const structure = indexFile.getStructure();

    expect(indexFile.getFilePath()).toEqual(`${getMockDir()}/index.ts`);
    expect(structure.statements).toHaveLength(1);
    expect((structure.statements as StatementStructures[])[0]).toEqual({
      kind: StructureKind.ExportDeclaration,
      moduleSpecifier: './test',
      isTypeOnly: false,
      namedExports: [],
      namespaceExport: undefined,
      attributes: undefined,
    });
    assertProjectSnapshot(project);
  });

  it('should return undefined if no files are generated', () => {
    const project = makeMockProject();
    const indexFile = namespaceGenerator({
      project,
      dir: getMockDir(),
      generator: () => [],
    });

    expect(indexFile).toBeUndefined();
    assertProjectSnapshot(project);
  });
});
