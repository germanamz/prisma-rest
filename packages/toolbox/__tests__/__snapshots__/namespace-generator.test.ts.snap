// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`namespaceGenerator should import the generated files correctly 1`] = `
"export * from "./test";
"
`;
